// helius-sdk/dist/examples/optimized_transactions/sendSmartTransactionWithWalletAdapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const src_1 = require("../../src"); // Replace with 'helius-sdk' in a production setting
const web3_js_1 = require("@solana/web3.js");
function main() {
    return __awaiter(this, void 0, void 0, function* () {
        const helius = new src_1.Helius("YOUR_API_KEY");
        // Simulated wallet adapter
        const mockWallet = web3_js_1.Keypair.generate(); // Replace with your keypair
        const payer = mockWallet.publicKey;
        const signTransaction = (tx) => __awaiter(this, void 0, void 0, function* () {
            if (tx instanceof web3_js_1.VersionedTransaction) {
                tx.sign([mockWallet]);
            }
            else {
                tx.partialSign(mockWallet);
            }
            return tx;
        });
        const recipient = new web3_js_1.PublicKey("TARGET_WALLET_ADDRESS"); // Replace this
        const instructions = [
            web3_js_1.SystemProgram.transfer({
                fromPubkey: payer,
                toPubkey: recipient,
                lamports: 0.01 * web3_js_1.LAMPORTS_PER_SOL,
            }),
        ];
        const lookupTables = [];
        const signature = yield helius.rpc.sendSmartTransactionWithWalletAdapter(instructions, payer, signTransaction, lookupTables, {
            priorityFeeCap: 1000000, // 0.001 SOL tip cap
            skipPreflight: true,
            pollTimeoutMs: 60000,
            pollIntervalMs: 2000,
            pollChunkMs: 10000,
            lastValidBlockHeightOffset: 150,
            preflightCommitment: "confirmed",
        });
        console.log(`Transaction confirmed: ${signature}`);
        console.log(`View on Orb: https://orb.helius.dev/tx/${signature}`);
    });
}
main().catch((err) => {
    console.error("Example failed:", err);
    process.exit(1);
});
