// helius-sdk/dist/examples/optimized_transactions/sendTransaction.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const src_1 = require("../../src"); // Replace with 'helius-sdk' in a production setting
const web3_js_1 = require("@solana/web3.js");
function main() {
    return __awaiter(this, void 0, void 0, function* () {
        const helius = new src_1.Helius("YOUR_API_KEY");
        const sender = web3_js_1.Keypair.generate(); // Replace with your loaded keypair
        const recipient = new web3_js_1.PublicKey("TARGET_WALLET_ADDRESS"); // Replace with recipient
        const { blockhash } = yield helius.connection.getLatestBlockhash();
        const transaction = new web3_js_1.Transaction({
            feePayer: sender.publicKey,
            recentBlockhash: blockhash,
        }).add(web3_js_1.SystemProgram.transfer({
            fromPubkey: sender.publicKey,
            toPubkey: recipient,
            lamports: 0.01 * web3_js_1.LAMPORTS_PER_SOL,
        }));
        transaction.sign(sender);
        const signature = yield helius.rpc.sendTransaction(transaction, {
            skipPreflight: true,
            preflightCommitment: "confirmed",
        });
        console.log(`Transaction sent! Signature: ${signature}`);
        console.log(`Explorer: https://orb.helius.dev/tx/${signature}`);
    });
}
main().catch((err) => {
    console.error("Example failed:", err);
    process.exit(1);
});
